docs/blueprints/linux-setup.md

```diff
@@ -0,0 +1,134 @@
+
+When building our package, we can execute  pre-installation and post-installation scripts. We could retrieve the storage node configuration (from command line for instance) from the user using one of these scripts.
+
+#### Agnostic Packaging
+There are [3 major agnostic packaging system](https://www.ostechnix.com/linux-package-managers-compared-appimage-vs-snap-vs-flatpak/) for linux: AppImage, FlatPak and Snap. As AppImage and FlatPak are more desktop application oriented, we choose to focus on Snap.
+
+##### Snap
+[Snaps](https://snapcraft.io/first-snap#go) are containerised software packages. They auto-update daily and work on a variety of Linux distributions. They also revert to the previous version if an update fails. This feature would make it necessary to find out how to implement the rollout versioning.
+
+From the [snap documentation](https://snapcraft.io/docs/go-applications), it seems pretty straightforward to package an application. Snaps are defined in a yaml file. Running an application as a service is done only by specifying "daemon: simple" in the application description. 
+This would make us save the work of building a storage node service.
+
+Snaps can then be published in the snapcraft [app store](https://snapcraft.io/). In the store, we would able to monitor the number of installed snaps. It is possible to [host our own store](https://ubuntu.com/blog/howto-host-your-own-snap-store) but that the snap daemon only handles one repository. Therefore, the use of Canonical's store seems mandatory. Snaps integrate well with [github](https://snapcraft.io/build).
+A snap inside the store can be published in multiple versions in different [channels](https://snapcraft.io/docs/channels).
+
+Snaps have been known for suffering a long start-up time, but it has been [improved](https://snapcraft.io/blog/snap-startup-time-improvements).
+
+#### Comparison between the package and the snap solutions
+Here is a table summarizing the differences between debian packaging system and snaps taken ([source](https://snapcraft.io/blog/a-technical-comparison-between-snaps-and-debs))
+
+| Package	| Debian	| Snap
+| --- | --- | ----
+| Format	| Ar archive	| SquashFS archive
+| Signature verification	| Y (often not used)	| Y
+| Package manager	| dpkg (low-level) Different higher-level  managers available	| snap
+| Front-end	| Many	| Snap Store
+| Installation	| Files copied to /	| Snap uncompressed and mounted as loopback device
+| Dependencies	| Shared	| Inside each snap or content snaps
+| Automatic updates	| semi-automatic	| Y
+| Transactional updates	| N	| Y
+| Multiple installs in parallel	| N	| Y
+| Multiple versions	| N	| Y
+| Security confinement	| Limited	| Y
+| Disk footprint	| Smaller	| Larger
+| Application startup time	| Default	| Typically longer
+
+We are thinking to use snap for the following reasons:
+- platform agnostic
+- support for daemons
+- auto-update with failure recovery
+One major drawback is Canonical central's repository.
```

Comment:
Snap is a reasonable choice. But here are some additional considerations:

* I don't think snaps are very popular on server-type machines. I think people tend to use snaps a lot more on the desktop and for GUI apps. (Although I don't know whether our Linux SNOs tend to run the software on a desktop or a headless box.) Users tend to prefer native packages more, when available.
* If we make debs and/or RPMs, we at least have a chance to be added to the applicable distros directly. We do have a way of getting a deb into the Debian main archive, and from there to Ubuntu, if our package quality meets standards (I am a Debian developer and think that Storj storage node software would be a great addition to both).
* Deb and RPM are similar enough in concept that the transition between them can be mostly automated, especially for packages like ours that don't need much interaction with the rest of the system. To my mind, this is sort of a "two for the price of 1" situation.
* I think we would get more support flowing back from the community for debs and RPMs, especially if we get added to the distros proper. Debian, Ubuntu, and Fedora contribute a lot back upstream, especially when needed to get well-behaved packages.
* Speaking to the points above,
  * Having a deb and an rpm is pretty close to supporting all linux target platforms as well. Snap is platform-agnostic in theory, but does require a working snapd, which not everyone has or wants to set up.
  * debs and rpms certainly support daemons.
  * debs and rpms do have failure recovery- there are facilities for putting things back the way they were when an upgrade fails. Debs in particular have a very well thought-out, battle-tested upgrade framework. Snaps can do this in a way that is closer to being atomic, but they have less control over how to fix state outside of the packaged software when an upgrade goes wrong.

Having said all that, I'll say again that Snap is a reasonable choice and the difference in outcome may be only marginal. If someone already has experience making snaps and is available to do it, that's almost certainly the easiest route for us.

Reply:
> Having a deb and an rpm is pretty close to supporting all linux target platforms as well. Snap is platform-agnostic in theory, but does require a working snapd, __which not everyone has or wants to set up__.

This was me; I've been reluctant to install snap in my distro and only I would do so if I don't find it in the native package manager and the installation is a nightmare or I have to manually touch more than a couple of files which has to be outside of my home directory.

Reply:
I think you're both right. I myself have been reluctant installing snap. We should use deb packaging. I'm modifying the document accordingly.